### ISSUE TITLE
[v2.7.7 Regression] Service "capi-webhook-service" not found

### ISSUE THREAD

**Issue by eliyamlevy** [rancher/rancher#42631]
**Rancher Server Setup**
- Rancher version: v2.7.7-rc4
- Installation option (Docker install/Helm Chart): Helm chart
- Kubernetes Version and Engine: v1.26.7-rancher1-1
- Backup Restore Version: v102.0.2+up3.1.2-rc1

**Describe the bug**
This issue is a regression cause by [this commit](https://github.com/rancher/rancher/commit/d3b5469238e965e54306f74d0e98a47b780cbe80) and more specifically this change
> Add rancher-provisioning-capi system-chart installation to replace embedded cluster-api controllers

What happens is that the restore is blocked because resources with the capi conversion webhooks are not able to find the capi-webhook-service. This is either because none of the capi resources are backed up or because of some kind of race condition that leads to the service not being ready when the resources are restored. 

**To Reproduce**
Steps to reproduce the behavior are basically the steps to do a migration:
1. Provision cluster with linked versions
2. Create one RKE and one RKE2 downstream clusters
3. Take backup of local cluster
4. Tear down local cluster
5. Restore on to properly configured fresh cluster

**Expected behavior**
Restore completes without errors

**Actual behavior**
Restore has below errors and constantly retrying without completion

**Log Output**
```
Error restoring namespaced resources [

  error restoring nick-277rc4-rke2 of type cluster.x-k8s.io/v1alpha4, Resource=clusters: restoreResource: err creating resource conversion webhook for cluster.x-k8s.io/v1alpha4, Kind=Cluster failed: Post "[https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s](https://capi-webhook-service.cattle-provisioning-capi-system.svc/convert?timeout=30s)": service "capi-webhook-service" not found 

  error restoring nick-277rc4-rke2-pool1 of type cluster.x-k8s.io/v1alpha4, Resource=machinedeployments: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=MachineDeployment failed: Post "[https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s](https://capi-webhook-service.cattle-provisioning-capi-system.svc/convert?timeout=30s)": service "capi-webhook-service" not found 

  error restoring nick-277rc4-rke2-pool1-7cd45d9cbcxbxzgj of type cluster.x-k8s.io/v1alpha4, Resource=machinesets: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=MachineSet failed: Post "[https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s](https://capi-webhook-service.cattle-provisioning-capi-system.svc/convert?timeout=30s)": service "capi-webhook-service" not found 

  error restoring nick-277rc4-rke2-pool1-7cd45d9cbcxbxzgj-7hg2h of type cluster.x-k8s.io/v1alpha4, Resource=machines: restoreResource: err creating resource conversion webhook for cluster.x-k8s.io/v1alpha4, Kind=Machine failed: Post "[https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s](https://capi-webhook-service.cattle-provisioning-capi-system.svc/convert?timeout=30s)": service "capi-webhook-service" not found 

  error restoring nick-277rc4-rke2-pool1-7cd45d9cbcxbxzgj-mmzmk of type cluster.x-k8s.io/v1alpha4, Resource=machines: restoreResource: err creating resource conversion webhook for cluster.x-k8s.io/v1alpha4, Kind=Machine failed: Post "[https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s](https://capi-webhook-service.cattle-provisioning-capi-system.svc/convert?timeout=30s)": service "capi-webhook-service" not found 

  error restoring nick-277rc4-rke2-pool1-7cd45d9cbcxbxzgj-hxn8c of type cluster.x-k8s.io/v1alpha4, Resource=machines: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=Machine failed: Post "[https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s](https://capi-webhook-service.cattle-provisioning-capi-system.svc/convert?timeout=30s)": service "capi-webhook-service" not found
]
```

**Additional context**
Tried doing a migration with `cluster.x-k8s.io/v1alpha4` not being backed up and it led to the restore successfully completing but I ran into issues with my RKE2 downstream cluster. One of the nodes was stuck in `waiting for machine plan to be applied` state.

Also tried doing a migration with the below resourceSelectors added in an attempt to back up capi-webhook and its accompanying service.
```
- apiVersion: "apps/v1"
  kindsRegexp: "^deployments$"  
  namespaces: 
    - "cattle-provisioning-capi-system"
- apiVersion: "apps/v1"
  kindsRegexp: "^services$"
  namespaces: 
    - "cattle-provisioning-capi-system"
```
The result was the same "service not found" error
```
ERRO[2023/09/01 17:12:21] Error restoring namespaced resources [
  error restoring rke2-ds of type cluster.x-k8s.io/v1alpha4, Resource=clusters: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=Cluster failed: Post "https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s": service "capi-webhook-service" not found 

  error restoring rke2-ds-pool1 of type cluster.x-k8s.io/v1alpha4, Resource=machinedeployments: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=MachineDeployment failed: Post "https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s": service "capi-webhook-service" not found 

  error restoring rke2-ds-pool1-cb98db4bx8p88d of type cluster.x-k8s.io/v1alpha4, Resource=machinesets: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=MachineSet failed: Post "https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s": service "capi-webhook-service" not found 

  error restoring rke2-ds-pool1-cb98db4bx8p88d-r9qkp of type cluster.x-k8s.io/v1alpha4, Resource=machines: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=Machine failed: Post "https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s": service "capi-webhook-service" not found 

  error restoring rke2-ds-pool1-cb98db4bx8p88d-fh2hd of type cluster.x-k8s.io/v1alpha4, Resource=machines: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=Machine failed: Post "https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s": service "capi-webhook-service" not found 

  error restoring rke2-ds-pool1-cb98db4bx8p88d-hdp5l of type cluster.x-k8s.io/v1alpha4, Resource=machines: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=Machine failed: Post "https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s": service "capi-webhook-service" not found
]
```

After trying this solution, but I believe the main problem to be a race condition. If this is the case it could take large changes to the restore controller code to fix.

**Comment by prachidamle** [comment]
@sowmyav27 we have not tested this with the already release backup-restore chart version  102.0.1+up3.1.1 yet, possibly it is a regression for already released backup-restore versions

**Comment by Oats87** [comment]
This is due to the newly introduced conversion webhooks that are a remnant of the fact that we bumped the `cluster-api` version in Rancher from `v1.2.x` to `v1.4.4`. We cannot easily backup/restore the cluster-api core controller manager deployment, as this will likely cause unintended behavior during restoration...

How do other object-level restoration solutions handle this?

**Comment by eliyamlevy** [comment]
Tried testing with rancher-backup version `102.0.1+up3.1.1` and got this error:

```
ERRO[2023/09/05 18:01:12] Error restoring namespaced resources [
  error restoring eli-rke2-ds of type cluster.x-k8s.io/v1alpha4, Resource=clusters: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=Cluster failed: Post "https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s": service "capi-webhook-service" not found 
  
  error restoring eli-rke2-ds-pool1 of type cluster.x-k8s.io/v1alpha4, Resource=machinedeployments: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=MachineDeployment failed: Post "https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s": service "capi-webhook-service" not found 
  
  error restoring eli-rke2-ds-pool1-6f675d9648xdz8s6 of type cluster.x-k8s.io/v1alpha4, Resource=machinesets: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=MachineSet failed: Post "https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s": service "capi-webhook-service" not found 
  
  error restoring eli-rke2-ds-pool1-6f675d9648xdz8s6-7xccv of type cluster.x-k8s.io/v1alpha4, Resource=machines: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=Machine failed: Post "https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s": service "capi-webhook-service" not found 
  
  error restoring eli-rke2-ds-pool1-6f675d9648xdz8s6-l7wql of type cluster.x-k8s.io/v1alpha4, Resource=machines: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=Machine failed: Post "https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s": service "capi-webhook-service" not found 
  
  error restoring eli-rke2-ds-pool1-6f675d9648xdz8s6-pdg8m of type cluster.x-k8s.io/v1alpha4, Resource=machines: restoreResource: err getting resource conversion webhook for cluster.x-k8s.io/v1beta1, Kind=Machine failed: Post "https://capi-webhook-service.cattle-provisioning-capi-system.svc:443/convert?timeout=30s": service "capi-webhook-service" not found
]
```
Looks like its broken with a previous version of backup restore and this result is to be expected.

**Comment by eliyamlevy** [comment]
I've done two migrations with the updated changes and the below versions. Everything was successful and no errors present. It does take some time to resolve the agent errors, so please let them work. It can take up to 15 minutes.

  Rancher: `v2.7.7-rc4`
  Backup Restore Operator: `v3.1.2-rc1` with `bundledeployments` and `cluster.x-k8s.io/v1alpha4` removed from resourceSet
  Kubernetes: `v1.26.7`

**Comment by prachidamle** [comment]
We should release-note this fix done and also this usecase about previous backup-restore operator releases mentioned here https://github.com/rancher/backup-restore-operator/pull/382#issuecomment-1710673295

**Comment by nickwsuse** [comment]
`Verified on v2.7-head ID: 60c50f0 && BRO v102.0.2+up3.1.2-rc2`

I retested all of the previous scenarios in [this ticket](https://github.com/rancher/rancher/issues/34518) plus a few extra since the capi-webhook change was included as well. There are no errors in the restore logs with these changes and restores are completed as expected. I did run into an issue of the resourceSet not being updated accordingly went upgrading Rancher and the BRO chart, logged [here](https://github.com/rancher/rancher/issues/42756). I also ran into  one of the migrations taking a while for the downstream clusters to come back up, but they did eventually come back up.

Test Cases:
1. Basic Migration
2. In-Place Restore
3. Rollback on same Rancher cluster
4. Upgrade Rancher and migrate
5. Rollback via migration
6. Backup without fleet → add fleet → restore to before fleet
7. The "more graceful" error state that was previously fixed in the first linked issue above
